[AMDGPU] Avoid changing minOccupancy if unclustered schedule was not run for any region.

dhruvachak opened the pull request:
During init of unclustered schedule stage, minOccupancy may be temporarily increased. But subsequently, if none of the regions are scheduled because they don't meet the conditions of initGCNRegion, minOccupancy remains incorrectly set. This patch avoids this incorrectness by delaying the change of minOccupancy until a region is about to be scheduled.

dhruvachak committed:
[AMDGPU] Reset minOccupancy if unclustered schedule was not run for any region.

During init of unclustered schedule stage, the minOccupancy is
temporarily increased. But subsequently, if none of the regions are
scheduled because they don't meet the conditions of initGCNRegion, the
minOccupancy should be reset to the initial occupancy. This change
detects the above situation and resets minOccupancy during finalization.

dhruvachak committed:
Move IsAnyRegionScheduled to UnclusteredHighRPStage.

dhruvachak committed:
Avoid modifying DAG and MFI minOccupancy before a region is scheduled.

arsenm commented on llvm/test/CodeGen/AMDGPU/schedule-regpressure-no-unclustered-regions.mir:
```diff
@@ -0,0 +1,735 @@
+# RUN: llc -mtriple=amdgcn -mcpu=gfx1100 -start-before=machine-scheduler -stop-after=greedy,2 -amdgpu-use-amdgpu-trackers=1 -verify-machineinstrs -debug-only=machine-scheduler %s -o - 2>&1 | FileCheck %s
```
```suggestion
# REQUIRES: asserts
# RUN: llc -mtriple=amdgcn -mcpu=gfx1100 -start-before=machine-scheduler -stop-after=greedy,2 -amdgpu-use-amdgpu-trackers -debug-only=machine-scheduler %s -o - 2>&1 | FileCheck %s
```

dhruvachak replied:
> 

Done.

dhruvachak committed:
Addressed feedback.

Reduced test case, added assert requirement, updated options.

dhruvachak committed:
Dropped all the LLT references.

dhruvachak commented on llvm/test/CodeGen/AMDGPU/schedule-regpressure-no-unclustered-regions.mir:
```diff
@@ -0,0 +1,735 @@
+    ret void
+  }
+
+  attributes #0 = { "amdgpu-flat-work-group-size"="1,256" }
```
I am not sure I can shrink the test using amdgpu-waves-per-eu. The test needs to have 2 things: 
```
MFI.getMaxWavesPerEU() > DAG.MinOccupancy
so that the target occupancy is incremented. This leads to the inconsistency in the current code. So I need this condition to be true in the test.
```
And the function RP has to be such that the region does not get scheduled. This is based on a number of conditions in UnclusteredHighRPStage::initGCNRegion. Given these conditions, I could not find a tighter range of amdgpu-waves-per-eu that allows shrinking the test. 

However, I was able to shrink the test significantly by just eliminating instructions as long as the above 2 conditions are met.

arsenm commented on llvm/test/CodeGen/AMDGPU/schedule-regpressure-no-unclustered-regions.mir:
```diff
@@ -0,0 +1,434 @@
+...
+
+# When using the GCN Trackers, the scheduler is able to acieve desired occupancy without running high-RP-reschedule stage. However, the RP is still high,
+# and RA is unable to allocate without spills. By running the high-RP-reschedule schedule we would have furhter decreased RP, which provides increased
```
```suggestion
# and RA is unable to allocate without spills. By running the high-RP-reschedule schedule we would have further decreased RP, which provides increased
```

arsenm reviewed (approved):
The test is still too big. I assume the scheduler respects RegisterClassInfo, does -stress-regalloc help?

dhruvachak committed:
Added option -stress-regalloc, allowing the test to be shortened.

Removed -amdgpu-use-amdgpu-trackers and associated comments since
they are not required any more for the checks.

Added one more CHECK line to capture both the conditions.

dhruvachak commented:
> The test is still too big. I assume the scheduler respects RegisterClassInfo, does -stress-regalloc help?

Yes, that helped a lot. The resulting test is much shorter.
